# Лабораторная работа №3

## Цель работы

Практическое применение библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.

## Задачи
1 Решить свой вариант диофантова уравнения.  
2 Имея открытую часть ключа RSA найти соответствующую закрытую часть и
расшифровать кодированное слово.  
3 Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим
выражением.

Вариант 3

## Решение диафантова уравнения

2601x − 3859y = 153

In [ ]:
from sympy.solvers.diophantine import diophantine
from sympy import symbols
x, y, z = symbols("x, y, z", integer=True)
diophantine(2601*x - 3859*y - 153)

{(227*t_0 + 414, 153*t_0 + 279)}

Ответ: (414, 279)

## Расшифровка слова

Русское слово из четырех букв закодировано при помощи алгоритма RSA открытым ключом (e = 15, m = 46).
Шифрованное сообщение имеет вид (28; 4; 37; 33). Подберите закрытую часть ключа и прочитайте исходное слово.
Буквам русского алфавита соответсвуют числа в диапазоне от 2 до 32 (исключены буквы “ё” и “ъ”).

In [ ]:
from sympy.crypto.crypto import decipher_rsa
from sympy.crypto.crypto import rsa_private_key
p, q, e = 23, 2, 15
prk = rsa_private_key(p, q, e)
print(decipher_rsa(28, prk), decipher_rsa(4, prk), decipher_rsa(37, prk), decipher_rsa(33, prk))

10 18 7 11


Закодированное слово:  ирей

## Запись булевой функции в ДНФ и КНФ

f(x, y, z) = ((y xor x)(xz)) ∨ (x xor (z ∨ y))

In [ ]:
from sympy.logic.boolalg import to_cnf, to_dnf
from sympy.abc import A, B, C
from sympy.logic.boolalg import Xor
print("CNF", to_cnf(((Xor(B, A)) & (A & C)) | (Xor(A, (C | B)))))
print("DNF", to_dnf(((Xor(B, A)) & (A & C)) | (Xor(A, (C | B)))))

CNF (A | B | C) & (~A | ~B) & (A | ~A | ~B) & (A | ~A | ~C) & (C | ~A | ~B) & (C | ~A | ~C) & (~A | ~B | ~C) & (A | B | ~A | ~B) & (A | B | ~A | ~C) & (A | B | C | ~A | ~B)
DNF (A & ~A) | (B & ~A) | (C & ~A) | (A & C & ~A) | (A & C & ~B) | (A & ~B & ~C) | (B & ~B & ~C) | (C & ~B & ~C) | (A & B & C & ~A) | (A & B & C & ~B)


# Выводы

В ходе выполнения работы были изучены и практически применены библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.